### ISSUE TITLE
Clear cache on product add/edit

### ISSUE THREAD

**Issue by leo72** [budgetneon/pagecache#14]
First of all this extension is great. Thank you very much!
I think the missing functionality is clearing cache on product create/edit/delete.
For example I have latest products list on the homepage and when I add new product it appears on homepage only after cache expires.

**Comment by budgetneon** [comment]
That's certainly a good idea.

We don't have the time to add that functionality at the moment, but I'll leave this open as a enhancement request.

**Comment by dhaupin** [comment]
Just tried this out on a test store, its pretty sweet, thanks. I agree though that per-entity purge is a make or break kinda thing....especially with a large store that is constantly being updated by 5 admin users. Otherwise, would be clearing the cache nonstop all day long during edit->check->edit->check->edit->pass...which obviously invalidates the purpose of a cache at all, and makes the OP's life tedious :)

Like 15 mins into looking at it, not sure how to do it yet considering things like module placement and multistores domains. Although you dont have time, any starting point thoughts on the individual purge tactic?

**Comment by budgetneon** [comment]
>>I agree though that per-entity purge is a make or break kinda thing

For some sites, perhaps.  There are lots of sites, ours included, running the cache without issue.  We don't typically make changes all day long though.  We'll do them at the end of the day, and clear the cache once.

A 100% clean per entity purge on many sites would be be pretty complex, as you mention.  We're caching pages, and with opencart, there's not a clear programmatic tie between product and "all pages a product might appear on".  There are a number of widgets and modules, including ones that ship with opencart, that display product pricing, stock levels, add to cart buttons, etc.

Another approach would be to zero on what exactly is changing, and just load those things via ajax.  For example, if it's prices that are constantly being updated, just load the price via ajax on page load, and overwrite the price in the cached page.

**Comment by dhaupin** [comment]
Aye its not the caching that runs with issue, its human imperfection that causes the issue. Not complaining though, this mod is cool. Just trying to explore how to pull off any sort of individual percolation. I realize mods are a lol thing to try to tackle, perhaps individual entities are not though.

Example: How often do you edit a page and get it right the first time? Nail a responsive grid flawless first WYSIWYG edit? Proof-read and don't need to change it? Guessing like the rest of the world it takes 2-3 saves to get something just right, which means 2-3 cache clears per entity as you proof them. That alone is applicable to *ALL* sites, otherwise, again, at the end of the day it takes multi cache clears to edit->check->edit, move to next to proof...makes for a long end of the day.

Looking at this more too -- about the AJAX stuff, seems to be caching those catalog side, at least in frontside real-time actions such as inline editing or chat-esque features (which doesnt seem like a big deal i guess, will just stack up one time use caches on 20 sec pings). Just figured ya should be aware of that too since it will make JSON reply cache files into the millions within a day on a large traffic'd site

**Comment by budgetneon** [comment]
Yeah, I'm not quite sure how to approach the problem since opencart doesn't provide an obvious way to map "editable thing" to "urls that changed as a result".

>>Looking at this more too -- about the AJAX stuff, seems to be caching those catalog side, at least in frontside real-time actions such as inline editing or chat-esque features 

If you're aware of any stock opencart AJAX that's being cached (and shouldn't be), let me know...I'll fix it.   

There is some logic there for both url patterns that shouldn't be cached, as well as some general situations (user is logged in, for example) where we won't cache.  If the cached ajax is a third party module, you would have to add entries to the $skip_urls array.  We do talk about that here: https://github.com/budgetneon/pagecache#caveats

There's also already a test to see if what's being cached is html or not, so it would be fairly easy to add a "only cache html" button.  Currently we use to test to avoid putting the html comment line into JSON.

**Comment by dhaupin** [comment]
Hah i see, I dunno why i didnt exclude that route (def was a 3rd party mod forgot was in there). Ill watch for stock ones though in the future, although i doubt there will be any like that. Thanks man

**Comment by budgetneon** [comment]
Sure.  FYI, the pattern is specifically a url, though...not a route.   It's matching on $_SERVER['REQUEST_URI'], like this:

    foreach ($this->skip_urls as $urlpattern) {
        if (preg_match($urlpattern,$_SERVER['REQUEST_URI'])) {
            $this->oktocache=false;
            return $this->oktocache;
        }
    }

**Comment by dhaupin** [comment]
Mod still kicking butt, thanks again budgetneon. Forgot to share this back, its a hella basic way to do this, and clears EVERYTHING, but it works pretty well. This will wipe cache after saving just about anything in the store that would affect the catalog side html generation (vQmod snippet for OC 1.5.x):

		<file path="admin/controller/" name="catalog/*.php,design/*.php,localisation/*.php,setting/*.php">

			<operation info="add cache clear for common strings in controllers" error="skip">
				<search position="after"><![CDATA[
					if (($this->request->server['REQUEST_METHOD'] == 'POST') && $this->validateForm()) {
				]]></search>
				<add><![CDATA[
				if (file_exists(DIR_SYSTEM . 'library/pagecache.php')) {
					require_once(DIR_SYSTEM . 'library/pagecache.php');
					$pagecache = new PageCache();
					$vals=$pagecache->Settings();
					$expire=$vals['expire'];
					$cachefolder=$vals['cachefolder'];
					$range=array( '0','1','2','3','4','5','6','7',
								  '8','9','a','b','c','d','e','f');
					$count=0;
					foreach ($range as $f) {
						foreach ($range as $s) {
							$dname=$cachefolder . $f . '/' . $s;
							if (is_dir($dname) && @$dir=opendir($dname)) {
								while (false !== ($file = readdir($dir))) {
								   // only purge files that end in .cache
								   if (substr($file,-6) == '.cache') {
									   $fpath=$dname . '/' . $file;
									   unlink($fpath);
									   $count++;
								   }
								}
							}
						}
					}
				}
				]]></add>
			</operation>
			
		</file>
		<file path="admin/controller/" name="module/*.php,payment/*.php,setting/*.php,shipping/*.php,total/*.php">

			<operation info="add cache clear for alternate common strings in controllers" error="skip">
				<search position="after"><![CDATA[
					if (($this->request->server['REQUEST_METHOD'] == 'POST') && $this->validate()) {
				]]></search>
				<add><![CDATA[
				if (file_exists(DIR_SYSTEM . 'library/pagecache.php')) {
					require_once(DIR_SYSTEM . 'library/pagecache.php');
					$pagecache = new PageCache();
					$vals=$pagecache->Settings();
					$expire=$vals['expire'];
					$cachefolder=$vals['cachefolder'];
					$range=array( '0','1','2','3','4','5','6','7',
								  '8','9','a','b','c','d','e','f');
					$count=0;
					foreach ($range as $f) {
						foreach ($range as $s) {
							$dname=$cachefolder . $f . '/' . $s;
							if (is_dir($dname) && @$dir=opendir($dname)) {
								while (false !== ($file = readdir($dir))) {
								   // only purge files that end in .cache
								   if (substr($file,-6) == '.cache') {
									   $fpath=$dname . '/' . $file;
									   unlink($fpath);
									   $count++;
								   }
								}
							}
						}
					}
				}
				]]></add>
			</operation>
			
		</file>

Also there is this, which shows the cache averages in the top bar, updates them every minute if anyone wants it (vQmod snippet for OC 1.5.x):

		<file name="admin/view/template/common/header.tpl">

			<operation info="add pagecache info in header" error="log">
				<search position="after"><![CDATA[
					<ul class="right" style="display: none;">
				]]></search>
				<add><![CDATA[
					<li><a href="<?php echo $this->url->link('module/pagecache', '&token=' . $this->session->data['token'], 'SSL'); ?>" class="pageCacheStage top" title="Cache File Count &amp; Size"></a></li>
	<script type="text/javascript"></script>
				]]></add>
			</operation>

		</file>

**Comment by IP-CAM** [comment]
Very need Job, dhaupin, I now just have to modify the other Mod a little, to make it work on OC v.1.5.6.5 as well! Thanks a LOT!

Ernie
![Uploading PageCacheView.jpg…]()

**Comment by IP-CAM** [comment]
Very need Job, dhaupin, I now just have to find out, how to make it work on my Test OC v.1.5.6.5 as well! Thanks a LOT!
Ernie - IP_CAM
![pagecacheview](https://cloud.githubusercontent.com/assets/12776547/8399982/ebba5880-1e17-11e5-81ee-cdcadb2499b4.jpg)
